presto-main/src/main/java/com/facebook/presto/cost/CachingStatsProvider.java

```diff
@@ -82,9 +102,129 @@ private PlanNodeStatsEstimate getGroupStats(GroupReference groupReference)
+                return null;
+            }
+
+            // Note: plans sharing the same CacheKey describe the same relation. However, calculated stats may differ,
+            // e.g. due to NDV capping when reducing JOIN is applied before a JOIN that multiplies rows.
+            // Here we ignore those differences for performance reasons.
+            CacheKey cacheKey = getCacheKey(node);
+            PlanNodeStatsEstimate stats = cache.get(cacheKey);
+            if (stats != null) {
+                return stats;
+            }
+            stats = statsProvider.apply(node);
+            verify(cache.put(cacheKey, stats) == null, "Stats already set");
+            return stats;
+        }
+
+        private CacheKey getCacheKey(PlanNode planNode)
+        {
+            CacheKeyBuilder builder = new CacheKeyBuilder();
+            planNode.accept(builder, null);
+            return builder.build();
+        }
+
+        private boolean isInnerJoinNode(PlanNode node)
+        {
+            if (!(node instanceof JoinNode)) {
+                return false;
+            }
+            JoinNode joinNode = (JoinNode) node;
+            return joinNode.getType() == INNER
+                    && (joinNode.getLeft() instanceof GroupReference || isInnerJoinNode(joinNode.getLeft()))
+                    && (joinNode.getRight() instanceof GroupReference || isInnerJoinNode(joinNode.getRight()));
+        }
+
+        private static final class CacheKeyBuilder
+                extends SimplePlanVisitor<Void>
+        {
+            private final ImmutableSet.Builder<Integer> groupIds = ImmutableSet.builder();
+            private final ImmutableSet.Builder<JoinNode.EquiJoinClause> criteria = ImmutableSet.builder();
+            private final ImmutableSet.Builder<Expression> filters = ImmutableSet.builder();
+
+            @Override
+            protected Void visitPlan(PlanNode node, Void context)
+            {
+                throw new UnsupportedOperationException("Unsupported node type: " + node.getClass());
+            }
+
+            @Override
+            public Void visitGroupReference(GroupReference node, Void context)
+            {
+                groupIds.add(node.getGroupId());
+                return null;
+            }
+
+            @Override
+            public Void visitJoin(JoinNode node, Void context)
+            {
+                super.visitPlan(node, context);
+                verify(node.getType() == INNER);
+                criteria.addAll(node.getCriteria());
+                node.getFilter().ifPresent(filters::add);
+                return null;
+            }
+
+            public CacheKey build()
+            {
+                return new CacheKey(groupIds.build(), criteria.build(), filters.build());
+            }
+        }
+
+        private static final class CacheKey
+        {
+            private final Set<Integer> groupIds;
+            private final Set<JoinNode.EquiJoinClause> criteria;
```

Comment:
I could be wrong, but it seems fragile. We may need to change the channel ids at some point.

Reply:
I don't understand. Could you please elaborate?

Reply:
Depending on the transformation we are doing with the joins the equivalent filter may be represented in a different manners. For example `channel1 = channel2` ~= `channle2 = channel1`. 

Or, if one day we change something in the join reordering code, the channels name might not exactly match, as we are technically allowed to do not preserve channel names. For example we are allowed to introduce additional projections, etc.

I'm not sure how likely, but it is possible to break this caching layer. It is not a big deal though. But since it is not a big deal (the caching saves us very little cpu time) - the question raises if we need it at all?

Reply:
You kind of convinced me not to introduce this caching layer. @kokosing what's your thinking?

Reply:
Sounds good!

Reply:
Also I think we should change the TODO comment that we should introduce this caching. And say there that it does not bring much benefit.